src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs

```diff
@@ -1010,77 +870,379 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+                {
+                    using (var sourceFile = new FileStream(file, FileMode.Open))
+                    {
+                        var sourceText = SourceText.From(sourceFile);
+                        _syntaxTrees.Add(VisualBasicSyntaxTree.ParseText(sourceText, _visualbasicParseOptions, file));
+                    }
+                }
+            }
+            else if (String.Equals(ParameterSetName, "FromMember", StringComparison.OrdinalIgnoreCase))
+            {
+                _sourceCode = GenerateTypeSource(Namespace, Name, _sourceCode, Language);
+
+                var sourceText = SourceText.From(_sourceCode);
+                _syntaxTrees.Add(VisualBasicSyntaxTree.ParseText(sourceText, _visualbasicParseOptions));
+            }
+            else if (String.Equals(ParameterSetName, "FromSource", StringComparison.OrdinalIgnoreCase))
+            {
+                var sourceText = SourceText.From(_sourceCode);
+                _syntaxTrees.Add(VisualBasicSyntaxTree.ParseText(sourceText, _visualbasicParseOptions));
+            }
+            else
+            {
+                Diagnostics.Assert(false, "Invalid parameter set: {0}", this.ParameterSetName);
+            }
+
+            if (CompileOnly)
+            {
+                VisualBasicCompileToAssembly();
+            }
+            else
+            {
+                // if the source code was already compiled and loaded and not changed
+                // we get the assembly from the cache.
+                if (isSourceCodeUpdated(out Assembly assembly))
+                {
+                    VisualBasicCompileToAssembly();
+                }
+                else
+                {
+                    if (PassThru)
+                    {
+                        WriteTypes(assembly);
+                    }
+
+                    WriteVerbose(AddTypeStrings.AlreadyCompiledandLoaded);
+                }
+            }
+        }
+
+        private bool isSourceCodeUpdated(out Assembly assembly)
+        {
+            Diagnostics.Assert(_syntaxTrees.Count != 0, "_sourceCode or _syntaxTrees should contains a source code.");
+
+            _syntaxTreesHash = ArraySyntaxTreeGetHashCode(_syntaxTrees);
+
+            if (s_sourceAssemblyCache.TryGetValue(_syntaxTreesHash, out Assembly hashedAssembly))
+            {
+                assembly = hashedAssembly;
+                return false;
+            }
+            else
+            {
+                assembly = null;
+                return true;
+            }
+        }
+
+        private void CSharpCompileToAssembly()
+        {
+            IEnumerable<PortableExecutableReference> references = GetPortableExecutableReferences();
+
+            Compilation compilation = CSharpCompilation.Create(
+                PathType.GetRandomFileName(),
+                syntaxTrees: _syntaxTrees,
+                references: references,
+                options: _csharpCompilationOptions);
```

Comment:
`_csharpCompilationOptions` can also be made a local variable in `CSharpSourceCodeProcessing()`, by changing this method to take a parameter of `CompilationOptions compilationOptions`.

Reply:
Fixed